### PR TITLE
Port of Javascript Solana library

### PR DESCRIPTION
**Pull request by exzachlyvv**
This PR contains a majority of the core Solana API classes and interactions. It leverages PHP's [sodium](https://www.php.net/manual/en/book.sodium.php) library for all things related to cryptography (TY sodium!).

I have done enough to submit real transactions successfully to the Solana blockchain (see SolanaRpcClientTest@it_test_zvv_real_transaction). Some areas have not been tested thoroughly yet.

Most of my reference point was the [1st party Javascript library](https://github.com/solana-labs/solana-web3.js), but also quite a bit from the community [python library](https://github.com/michaelhly/solana-py) and a little from the community [java library](https://github.com/skynetcapital/solanaj).

Some helpful things to know:
1. In other languages, binary data is typically an array of bytes. In PHP, binary is typically a string (b"asdf"). There are 2 nice helpers in this PR, `Ed25519Keypair::array2bin()` and `Ed25519Keypair::bin2array()`

TODO probably in a different PR (in no particular order):
1. Abstract away the binary implications of TransactionInstruction creation (see SystemProgram@transfer()). I hope developers do not need to be concerned with anything binary-related when using this package.
2. Now that I better understand many of the moving parts, I'm sure there are many improvements to make.
3. Build out more of the native API calls in the Connection class.
4. Build out SystemProgram, SlpTokenProgram, StakeProgram (?), MetaplexProgram (?)
5. Add linting/cs-fixer.
6. Specific Exceptions. Right now I use the GenericException all over the place
7. code coverage + unit test missing areas.

### CODE REVIEW

**Comment by exzachlyvv** [comment]
Please do, go crazy 😄

File: src/Account.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                ? $secretKey
+                : Ed25519Keypair::array2bin($secretKey);
+
+            $this->keyPair = KeyPair::fromSecretKey($secretKeyString);
+        } else {
+            $this->keyPair = new KeyPair();
+        }
+    }
+
+    /**
```
Comment (mattstauffer) [inline src/Account.php:24]: I usually drop the doc blocks when we're using return types. Any reason to keep them, though?

File: src/Connection.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace Tighten\SolanaPhpSdk;
+
+use Tighten\SolanaPhpSdk\Exceptions\AccountNotFoundException;
+use Tighten\SolanaPhpSdk\Util\Commitment;
+use Tighten\SolanaPhpSdk\Util\Ed25519Keypair;
+
+class Connection extends Program
+{
+    /**
```
Comment (mattstauffer) [inline src/Connection.php:11]: Same note RE doc blocks for all of these; if we're type hinting I think we can drop the doc blocks unless they provide any value you know of?

File: src/Connection.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+     * @param string $transactionSignature
+     * @return array
+     */
+    public function getConfirmedTransaction(string $transactionSignature): array
+    {
+        return $this->client->call('getConfirmedTransaction', [$transactionSignature]);
+    }
+
+    /**
+     * NEW: This method is only available in solana-core v1.7 or newer. Please use getConfirmedTransaction for solana-core v1.6
```
Comment (mattstauffer) [inline src/Connection.php:45]: This sort of thing I would definitely still keep in the doc block though.

File: src/Connection.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * @param Transaction $transaction
+     * @param KeyPair $signer
+     * @param array $params
+     * @return array|\Illuminate\Http\Client\Response
+     * @throws Exceptions\GenericException
+     * @throws Exceptions\InvalidIdResponseException
+     * @throws Exceptions\MethodNotFoundException
+     */
+    public function sendTransaction(Transaction $transaction, KeyPair $signer, $params = [])
```
Comment (mattstauffer) [inline src/Connection.php:76]: I think we can use union types here, right? and then we can do a return type and drop the docblock? I don't mind keeping the docblock for the throws though.

File: src/Account.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Tighten\SolanaPhpSdk;
+
+use Tighten\SolanaPhpSdk\Util\Ed25519Keypair;
+
+class Account
+{
+    protected KeyPair $keyPair;
+
+    public function __construct($secretKey = null)
```
Comment (mattstauffer) [inline src/Account.php:11]: I'm ok if we wanna do a union type on `$secretKey`

File: src/Exceptions/TodoException.php [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+namespace Tighten\SolanaPhpSdk\Exceptions;
+
+use Exception;
+use Throwable;
+
+class TodoException extends Exception
+{
+    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    {
+        parent::__construct($message . " Help is appreciated: https://github.com/tighten/solana-php-sdk", $code, $previous);
```
Comment (mattstauffer) [inline src/Exceptions/TodoException.php:12]: Love it!

```suggestion
        parent::__construct($message . " | Help is appreciated: https://github.com/tighten/solana-php-sdk", $code, $previous);
```

File: src/KeyPair.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * An account keypair used for signing transactions.
+ */
+class KeyPair
+{
+    protected Ed25519Keypair $_keypair;
+
+    /**
+     * @param $secretKey
+     */
```
Comment (mattstauffer) [inline src/KeyPair.php:17]: Wrong docblock--why I love dropping docblocks when they're not needed :D

File: src/KeyPair.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    public function __construct(?Ed25519Keypair $keypair = null)
+    {
+        $this->_keypair = $keypair ?? Ed25519Keypair::generate();
+    }
+
+    /**
+     * Generate a new random keypair
+     */
+    static function generate(): KeyPair
```
Comment (mattstauffer) [inline src/KeyPair.php:26]: I *think* you can type `static`.. not a huge difference but I think it's a bit clearer.

```suggestion
    static function generate(): static
```

File: src/KeyPair.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+     * Create a keypair from a raw secret key byte array.
+     *
+     * This method should only be used to recreate a keypair from a previously
+     * generated secret key. Generating keypairs from a random seed should be done
+     * with the {@link Keypair.fromSeed} method.
+     *
+     * @param $secretKey
+     * @return KeyPair
+     */
+    static public function fromSecretKey(string $secretKey): KeyPair
```
Comment (mattstauffer) [inline src/KeyPair.php:41]: Same note here re returning `static`

Comment (mattstauffer) [reply src/KeyPair.php:41]: (and all the rest)

File: src/Message.php [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    )
+    {
+        $this->header = $header;
+        $this->accountKeys = array_map(function (string $accountKey) {
+            return new PublicKey($accountKey);
+        }, $accountKeys);
+        $this->recentBlockhash = $recentBlockhash;
+        $this->instructions = $instructions;
+
+        $this->indexToProgramIds = [];
```
Comment (mattstauffer) [inline src/Message.php:52]:
```suggestion
        $this->indexToProgramIds = [];
        
```

File: src/Message.php [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+        return Ed25519Keypair::array2bin($out);
+    }
+
+    /**
+     * @return array
+     */
+    protected function encodeMessage(): array
+    {
+        $publicKeys = [];
```
Comment (mattstauffer) [inline src/Message.php:127]:
```suggestion
        $publicKeys = [];
        
```

File: src/Message.php [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        return new Message(
+            $header,
+            $accountKeys,
+            $recentBlockhash,
+            $instructions
+        );
+    }
+}
```
Comment (mattstauffer) [inline src/Message.php:218]: Bro. I'm so glad you're the one who brought this class over and not me 😅

File: src/Programs/MetaplexProgram.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+{
+    public const METAPLEX_PROGRAM_ID = 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s';
+
+    /**
+     * @param $client
+     */
+    public function __construct($client)
+    {
+        parent::__construct($client);
+    }
```
Comment (mattstauffer) [inline src/Programs/MetaplexProgram.php:17]: Unnecessary constructor

File: src/Programs/SplTokenProgram.php [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public const SOLANA_TOKEN_PROGRAM_ID = 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA';
+
+    /**
+     * @param $client
+     */
+    public function __construct($client)
+    {
+        parent::__construct($client);
+    }
```
Comment (mattstauffer) [inline src/Programs/SplTokenProgram.php:17]: Unnecessary constructor

File: src/Programs/SystemProgram.php [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+use Tighten\SolanaPhpSdk\PublicKey;
+use Tighten\SolanaPhpSdk\TransactionInstruction;
+use Tighten\SolanaPhpSdk\Util\AccountMeta;
+
+class SystemProgram extends Program
+{
+    const PROGRAM_INDEX_CREATE_ACCOUNT = 0;
+    const PROGRAM_INDEX_TRANSFER = 2;
+
+    /**
```
Comment (mattstauffer) [inline src/Programs/SystemProgram.php:17]: Unnecessary constructor

File: src/PublicKey.php [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+    /**
+     * @return PublicKey
+     */
+    public static function default(): PublicKey
+    {
+        return new static('11111111111111111111111111111111');
+    }
+
+    /**
+     * Checks if two publicKeys are equal
```
Comment (mattstauffer) [inline src/PublicKey.php:52]:
```suggestion
     * Check if two publicKeys are equal
```

File: src/PublicKey.php [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+    static function isOnCurve($publicKey): bool
+    {
+        try {
+            $binaryString = $publicKey instanceof PublicKey
+                ? $publicKey->toBinaryString()
+                : $publicKey;
+
+            $_ = sodium_crypto_sign_ed25519_pk_to_curve25519($binaryString);
+            return true;
+        } catch (\SodiumException $exception) {
```
Comment (mattstauffer) [inline src/PublicKey.php:192]: Would you mind importing this and the Exception above on line 170?

File: src/Transaction.php [hunk 18]
```diff
@@ -0,0 +1,655 @@
...
+    {
+        if ($source instanceof KeyPair) {
+            return Ed25519Keypair::array2bin($source->getSecretKey());
+        } elseif ($source instanceof Signer) {
+            return $source->secretKey;
+        } else {
+            throw new GenericException('Invalid $needle input into arraySearchAccountMetaForPublicKey');
+        }
+    }
+}
```
Comment (mattstauffer) [inline src/Transaction.php:655]: Again.. glad this class was you and not me. I'm so impressed with how much work you've put into this!

File: src/Message.php [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+     * @param array<string> $accountKeys
+     * @param string $recentBlockhash
+     * @param array<CompiledInstruction> $instructions
+     */
+    public function __construct(
+        MessageHeader $header,
+        array         $accountKeys,
+        string        $recentBlockhash,
+        array         $instructions
+    )
```
Comment (mattstauffer) [inline src/Message.php:43]: For consistency can we drop the space alignment here and let them fall ragged?

**Review by mattstauffer — CHANGES_REQUESTED** [review]
OK, so.. I got a third of the way into this review and it's *clear* to me you have a much greater capacity and knowledge and amount of time to work on this than me.

I want to *fully* offer to drop the Tighten version of this package and let you take it over under your namespace. Let me know what you think!

File: src/Account.php (hunk 1 above)
Comment (exzachlyvv) [reply src/Account.php:24]: I know this is pretty debated, but I LOVE the doc blocks for the visual separation, even if their contents add no value. I would enjoy leaving them in :)

File: src/Connection.php (hunk 2 above)
Comment (exzachlyvv) [reply src/Connection.php:11]: Visual separation 🤩

File: src/KeyPair.php (hunk 7 above)
Comment (exzachlyvv) [reply src/KeyPair.php:17]: 😂 - let me sleep on them please. I do LOVE them for the visual. We can remove them later if they prove to be a pain in the butt to keep in sync

File: src/KeyPair.php (hunk 8 above)
Comment (exzachlyvv) [reply src/KeyPair.php:26]: Love it!! Agreed cleaner, however, that is supported in PHP 8.0+. We probably should keep the min as low as possible for now?

**Comment by exzachlyvv** [comment]
From a functional standpoint, I am now ready to merge this.

**Comment by mattstauffer** [comment]
@exzachlyvv Thank you for your kind words!

And yah, if you'd rather be the maintainer but under the Tighten namespace, that's fine by me! Let me give you maintain permissions right now :)

File: src/Account.php (hunk 1 above)
Comment (mattstauffer) [reply src/Account.php:27]: :) Fine -- let's just make sure they are kept up to date please.

File: src/Transaction.php [hunk 20]
```diff
@@ -0,0 +1,650 @@
...
+                $numReadonlySignedAccounts,
+                $numReadonlyUnsignedAccounts
+            ),
+            $accountKeys,
+            $recentBlockhash,
+            $instructions
+        );
+    }
+
+    /**
```
Comment (mattstauffer) [inline src/Transaction.php:278]: Can we leave behind a todo or an issue marking why this is commented out? Whether it's something we want to implement but haven't, or whatever else?

**Review by mattstauffer — CHANGES_REQUESTED** [review]
Freaking fantastic!

Only thing beyond these notes: Could you add your name to the credits on the readme and to composer.json as an author? Thanks!

**Comment by mattstauffer** [comment]
Great work! I'm gonna tag this as a new dot release so I can keep my old apps on the original until I update.

**Pull request merged**